x-pack/plugins/task_manager/task_manager.ts

```diff
@@ -0,0 +1,170 @@
+ * - post-init, it disallows middleware registration, but allows task manipulation
+ *
+ * Due to its complexity, this is mostly tested by integration tests (see readme).
+ */
+
+/**
+ * The public interface into the task manager system.
+ */
+export class TaskManager {
+  private isInitialized = false;
+  private maxWorkers: number;
+  private overrideNumWorkers: { [taskType: string]: number };
+  private definitions: TaskDictionary<SanitizedTaskDefinition>;
+  private store?: TaskStore;
+  private poller?: TaskPoller;
+  private middleware = {
+    beforeSave: async (saveOpts: BeforeSaveMiddlewareParams) => saveOpts,
+    beforeRun: async (runOpts: RunContext) => runOpts,
+  };
+
+  /**
+   * Initializes the task manager, preventing any further addition of middleware,
+   * enabling the task manipulation methods, and beginning the background polling
+   * mechanism.
+   */
+  public constructor(kbnServer: any, server: any, config: any) {
+    this.maxWorkers = config.get('xpack.task_manager.max_workers');
+    this.overrideNumWorkers = config.get('xpack.task_manager.override_num_workers');
+    this.definitions = {};
+
+    const logger = new TaskManagerLogger((...args: any[]) => server.log(...args));
+
+    kbnServer.afterPluginsInit(() => {
+      const store = new TaskStore({
+        callCluster: server.plugins.elasticsearch.getCluster('admin').callWithInternalUser,
+        index: config.get('xpack.task_manager.index'),
+        maxAttempts: config.get('xpack.task_manager.max_attempts'),
+        supportedTypes: Object.keys(this.definitions),
+      });
+      const pool = new TaskPool({
+        logger,
+        maxWorkers: this.maxWorkers,
+      });
+      const createRunner = (instance: ConcreteTaskInstance) =>
+        new TaskManagerRunner({
+          logger,
+          kbnServer,
+          instance,
+          store,
+          definition: this.definitions[instance.taskType],
+          beforeRun: this.middleware.beforeRun,
+        });
+      const poller = new TaskPoller({
+        logger,
+        pollInterval: config.get('xpack.task_manager.poll_interval'),
+        work() {
+          return fillPool(pool.run, store.fetchAvailableTasks, createRunner);
+        },
+      });
+
+      server.plugins.elasticsearch.status.on('red', () => {
+        logger.debug('Lost connection to Elasticsearch, stopping the poller.');
+        poller.stop();
+      });
+
+      server.plugins.elasticsearch.status.on('green', async () => {
+        logger.debug('Initializing store');
+        await store.init();
+        logger.debug('Starting poller');
+        await poller.start();
+        logger.info('Connected to Elasticsearch, and watching for tasks');
+      });
+
+      this.store = store;
+      this.poller = poller;
+      this.isInitialized = true;
+    });
+  }
+
+  public registerTaskDefinitions(taskDefinitions: TaskDictionary<TaskDefinition>) {
+    const sanitized = sanitizeTaskDefinitions(
```

Comment:
We should probably do an `assertUninitialized` here, too. I think having the task definitions mutate on us while running is a bad idea.